connectors/src/connectors/microsoft/lib/graph_api.ts

```diff
@@ -138,29 +142,115 @@ export async function getItem(client: Client, itemApiPath: string) {
   return client.api(itemApiPath).get();
 }
 
-export type MicrosoftNodeData = {
-  nodeType: MicrosoftNodeType;
-  itemApiPath: string;
+type MicrosoftEntity = {
+  folder: MicrosoftGraph.DriveItem;
+  drive: MicrosoftGraph.Drive;
+  site: MicrosoftGraph.Site;
+  team: MicrosoftGraph.Team;
+  file: MicrosoftGraph.DriveItem;
+  page: MicrosoftGraph.DriveItem;
+  channel: MicrosoftGraph.Channel;
+  message: MicrosoftGraph.ChatMessage;
+  worksheet: MicrosoftGraph.WorkbookWorksheet;
+};
+
+export type MicrosoftEntityMapping = {
+  [K in keyof MicrosoftEntity]: MicrosoftEntity[K];
 };
 
-export function microsoftInternalIdFromNodeData(nodeData: MicrosoftNodeData) {
+/**
+ * Converts a Microsoft entity to a MicrosoftNode depending on the nodeType
+ * Note: parentItemAPIPath is not set in the returned object, the parent logic
+ * is not explicit in the Microsoft Graph API and is handled on our side
+ * @param nodeType
+ * @param itemRaw
+ */
+export function itemToMicrosoftNode<T extends keyof MicrosoftEntityMapping>(
+  nodeType: T,
+  itemRaw: MicrosoftEntityMapping[T]
+): MicrosoftNode {
+  switch (nodeType) {
+    case "folder": {
+      const item = itemRaw as MicrosoftGraph.DriveItem;
```

Comment:
🙈 Did not get a chance to get there, but fine to unblock. I'd say that the simplest way is probably to change `itemRaw` from a conditional type to an object property that combine both or you can use assert.

Reply:
:+1: will try a few things